### DISCUSSION TITLE
Which code implement "reconciliation" in tw5?

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#7558] [Q&A] [answered]
> [What is reconciliation](https://github.com/acdlite/react-fiber-architecture#what-is-reconciliation)?
> 
> reconciliation
>     The algorithm React uses to diff one tree with another to determine which parts need to be changed.
> update
>     A change in the data used to render a React app. Usually the result of `setState`. Eventually results in a re-render. 

I thought tw's v-dom has this feature, but in my plugins, I directly create new DOM element in the `render` method, and append them to the parentElement provided by tw5.

After reading [need-for-speed thread in the talk forum](https://talk.tiddlywiki.org/t/the-5-000-tiddler-club-people-with-the-need-for-speed/4040/22), I start rethinking if I was doing wrong.

Or maybe tw5 just doesn't have this comparison? Thus every DOM gets updated when a tiddler (even a draft) changes. Maybe I have to read the code to find out.

**Comment by Jermolene** [comment, marked as answer]
TiddlyWiki doesn't use a virtual DOM. Instead, each widget is responsible for refreshing itself and any attached DOM nodes when there have been changes to the tiddler store. In other words, we refer to "reconciliation" as "refreshing".

**Reply by linonetwo** [reply]
Thanks, this is reasonable, because Widgets are not "pure functions", so we can't predict the output from its input. (If so, we can easily add some performance improvements).

Basically, "attribute" + "state" => output , we can know whether a widget's attributes changed, but currently many widgets put state on `this`, for example:

```ts
this.list = this.getTiddlerList();
```

so we can't know if it is changed, so we can't do the same thing as react.

**Reply by pmario** [reply]
The state of TW is kept in tiddlers. So if tiddlers are changed, the wiki-store `$tw.wiki` starts a `.dispatchEvent("change",changes)` event.. Every element, that has registered an eventHandler with `.wiki.addEventListener("change",function())` is notified. 

See: https://github.com/Jermolene/TiddlyWiki5/blob/a05302da10de8404ca5aff793c661260673c4dfd/core/modules/wiki.js#L145

The widget "refresh" mechanism is started with a "change" event. Some "early" eventListeners are registered at startup-time in **render.js**
https://github.com/Jermolene/TiddlyWiki5/blob/a05302da10de8404ca5aff793c661260673c4dfd/core/modules/startup/render.js#L30-L40

The first event that is triggered there is the one registered by `$tw.titleWidgetNode()`, which sets the `document.title` of the browser window. So if you change the content of the `$:/SiteTitle` in the [ControlPanel](https://tiddlywiki.com/#%24%3A%2FControlPanel), the browser document.title is updated immediately. So this info can be seen in the **browser tabs**. 

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/105bf373-987b-4b47-9415-a487c2436bae)

This event is easy to follow. Open [$:/core/wiki/title](https://tiddlywiki.com/#%24%3A%2Fcore%2Fwiki%2Ftitle) it contains 2 transclusions and 1 list-widget

```
{{$:/SiteTitle}}
<$list filter="[{$:/SiteSubtitle}trim[]minlength[1]]" variable="ignore">
&#32;---&#32;{{$:/SiteSubtitle}}
</$list>
```
Since the `$:/SiteSubtitle` is **not** changed, the subtitle transclusion widget should not be redrawn if you only change the `$:/SiteTitle`. 

**This can easily be tested with eg: Edge or Chrome Dev-Console -> Rendering tab -> Paint flashing enabled.** 

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/6a36af6a-3b85-4d30-9aa6-cccedf3b0203)

The transclusion `.refresh()` code can be found at: https://github.com/Jermolene/TiddlyWiki5/blob/a05302da10de8404ca5aff793c661260673c4dfd/core/modules/widgets/transclude.js#L406

There are several other elements that are updated: 

- [Title of this TiddlyWiki](https://tiddlywiki.com/#%24%3A%2FSiteTitle) - link, because it contains a link to the $:/SiteTitle tiddler. The link-widget refresh mechanism checks if there have been changes to one of it's attributes. The **to**-attribute `<$link to="$:/SiteTitle"><<lingo Title/Prompt>></$link>`

So if you have a look to the link-widget code: https://github.com/Jermolene/TiddlyWiki5/blob/a05302da10de8404ca5aff793c661260673c4dfd/core/modules/widgets/link.js#L208

- You can see, which widget-attributes are checked for a change that will cause the link-widget to destroy and re-render itself with `this.refreshSelf();`

- `return this.refreshChildren(changedTiddlers);` **will always be called** by every widget to inform their children to also check their refresh-state. 

----

At [Talk](https://talk.tiddlywiki.org/t/the-5-000-tiddler-club-people-with-the-need-for-speed/4040/22?u=pmario) you ask

> 1. Why every DOM get updated when a tiddler (even a draft) changes? ..

As you see above .. not every DOM element is redrawn, only widgets where related widget-attributes have been changed. So it's highly selective. Those changes are calculated in the `.render()` function with: `this.computeAttributes();` which is basically always one the first lines in any widget render() function. 

**IMO It should be similar** in your own .render() functions to be performant.

Hope that makes sense. 
-m

**Comment by pmario** [comment]
@linonetwo you wrote 

>https://github.com/Jermolene/TiddlyWiki5/discussions/7558#discussioncomment-6254441
>
> ```
> this.list = this.getTiddlerList();
> ```
> 
> so we can't know if it is changed, so we can't do the same thing as react.

It would be interesting to see the **context**. As I wrote above, TW state should always be kept in tiddlers. That's the only way to know when to refresh. ... I think there is a high chance, that the code above is there for easy use of the `this.list` for "local functions" which need access to the list in a closure.

**Reply by pmario** [reply]
I think I found the important context for the internal state. And the `this.list` is used very often in the code so it would not make sense to [recreate the list](https://github.com/Jermolene/TiddlyWiki5/blob/a05302da10de8404ca5aff793c661260673c4dfd/core/modules/widgets/list.js#L88) everytime the values are needed. 

https://github.com/Jermolene/TiddlyWiki5/blob/a05302da10de8404ca5aff793c661260673c4dfd/core/modules/widgets/list.js#L199-L202
